File: apps/web-map-console/src/lib/observability/sentry.ts
/**
 * Optional Sentry bootstrap untuk frontend.
 *
 * `@sentry/react` adalah optional dependency. Kalau modul tidak terinstall,
 * fungsi `initSentry` no-op sehingga build/dev tidak terganggu.
 *
 * Untuk mengaktifkan:
 *   npm install @sentry/react
 *   set VITE_SENTRY_DSN di .env.production
 */

export interface SentryBootstrapOptions {
  dsn?: string;
  environment?: string;
  tracesSampleRate?: number;
  release?: string;
}

type SentryReactModule = {
  init: (options: {
    dsn: string;
    environment: string;
    tracesSampleRate: number;
    release: string;
  }) => void;
};

const optionalImport = new Function("specifier", "return import(specifier)") as (
  specifier: string
) => Promise<SentryReactModule>;

export async function initSentry(options: SentryBootstrapOptions = {}): Promise<boolean> {
  const dsn = options.dsn ?? import.meta.env.VITE_SENTRY_DSN;
  if (!dsn) return false;
  try {
    const Sentry = await optionalImport("@sentry/react");
    Sentry.init({
      dsn,
      environment: options.environment ?? import.meta.env.VITE_SENTRY_ENVIRONMENT ?? "production",
      tracesSampleRate:
        options.tracesSampleRate ??
        Number(import.meta.env.VITE_SENTRY_TRACES_SAMPLE_RATE ?? 0.05),
      release: options.release ?? `time-rally-web@${import.meta.env.MODE}`,
    });
    return true;
  } catch {
    return false;
  }
}

File: services/api/app/observability/sentry.py
"""Optional Sentry initialization.

Sentry SDK adalah optional dependency. Kalau modul belum terinstal, fungsi
``init_sentry`` no-op sehingga test/dev tidak terganggu.

Untuk mengaktifkan: `pip install sentry-sdk` dan set env SENTRY_DSN.
"""

from __future__ import annotations

from app.settings import get_settings


def init_sentry() -> bool:
    settings = get_settings()
    if not settings.sentry_dsn:
        return False
    try:
        import sentry_sdk  # type: ignore[import-not-found]
        from sentry_sdk.integrations.fastapi import FastApiIntegration  # type: ignore[import-not-found]
        from sentry_sdk.integrations.starlette import StarletteIntegration  # type: ignore[import-not-found]
    except Exception:  # noqa: BLE001
        return False

    sentry_sdk.init(
        dsn=settings.sentry_dsn,
        environment=settings.sentry_environment,
        traces_sample_rate=settings.sentry_traces_sample_rate,
        integrations=[
            FastApiIntegration(transaction_style="endpoint"),
            StarletteIntegration(transaction_style="endpoint"),
        ],
        release=f"time-rally-api@{settings.app_env}",
    )
    return True
